### ISSUE TITLE
Should I Complete fixing for `MD013`?

### ISSUE THREAD

**Issue by n8behavior** [rvben/rumdl#13]
It seems that line wrapping, `MD013`, is not "fixable" yet?? 

![Image](https://github.com/user-attachments/assets/c9d514a3-aec7-4267-94a4-2cc478fc9675)

Guessing this is the problem, assuming I'm reading this correctly :)

https://github.com/rvben/rumdl/blob/15bb69f50db6ec2b774627ad80241ec193df69b8/src/rules/md013_line_length.rs#L245

Any reason I shouldn't add this and submit a PR? Just want to make sure I'm not duplicating work or missing something and this feature actually should work already ;)

[grumble] I'm tired of `markdownlint-cli2` in my default LazyVim config taking forever to format large docs, so I'm motivated to fix this [/grumble]

**Comment by rvben** [comment]
Hi @n8behavior , I had a version but I wasn't sure how the auto-fix should work. I was concerned it would lead to unexpected behavior for users. If you have a good suggestion/PR, I'd definitely consider it. If risky in some way, we can always choose to make it easily configurable.

**Comment by n8behavior** [comment]
Okay. I'm toying with an idea using https://github.com/pulldown-cmark/pulldown-cmark 

I kinda want to see if the event stream coming from a pull parser makes reflowing text easier. For example, the `MD013` rule would be interested in paragraph events and will want to reflow `Text` nodes while preserving others, e.g. `Link`.

I'm only kicking this around because I'm familiar with this parser, though fairly certain I can adapt to the current parser if needed. Anyway, consider this markdown (backticks replaced with single quotes so it formats correctly in this issue :)

```markdown
# Example Heading

Example paragraph with with a long sentence that needs wrap that also contains `code that maybe could wrap` as well. However, a [link](#) should never be split during text reflow!

This block must not wrap

'''Rust
for event in Parser::new(markdown_input) {
    match &event {
        ...
    }
}
'''

That's all, folks!
```

Here is some code to print the event stream

```Rust
...
    for event in Parser::new(markdown_input) {
        match &event {
            Event::Start(tag) => println!("Start: {:?}", tag),
            Event::End(tag) => println!("End: {:?}", tag),
            Event::Html(s) => println!("Html: {:?}", s),
            Event::InlineHtml(s) => println!("InlineHtml: {:?}", s),
            Event::Text(s) => println!("Text: {:?}", s),
            Event::Code(s) => println!("Code: {:?}", s),
            Event::DisplayMath(s) => println!("DisplayMath: {:?}", s),
            Event::InlineMath(s) => println!("Math: {:?}", s),
            Event::FootnoteReference(s) => println!("FootnoteReference: {:?}", s),
            Event::TaskListMarker(b) => println!("TaskListMarker: {:?}", b),
            Event::SoftBreak => println!("SoftBreak"),
            Event::HardBreak => println!("HardBreak"),
            Event::Rule => println!("Rule"),
        };
    }
...
```

And the event stream looks like...

```js
Start: Heading { level: H1, id: None, classes: [], attrs: [] }
Text: Borrowed("Example Heading")
End: Heading(H1)
Start: Paragraph
Text: Borrowed("Example paragraph with with a long sentence that needs wrap that also contains ")
Code: Borrowed("code that maybe could wrap")
Text: Borrowed(" as well. However, a ")
Start: Link { link_type: Inline, dest_url: Borrowed("#"), title: Borrowed(""), id: Borrowed("") }
Text: Borrowed("link")
End: Link
Text: Borrowed(" should never be split due during text reflow!")
End: Paragraph
Start: Paragraph
Text: Borrowed("This block must not wrap")
End: Paragraph
Start: CodeBlock(Fenced(Borrowed("Rust")))
Text: Borrowed("for event in Parser::new(markdown_input) {\n    match &event {\n        ...\n    }\n}\n")
End: CodeBlock
Start: Paragraph
Text: Borrowed("That's all, folks!")
End: Paragraph
```

Looking at those events, I'm thinking I could decide when/where to reflow (wrap) text safely in a single pass without having to walk a node tree or lookup context.  And it's pretty fast. A reasonably sized README.md, for example, streams through in a few milliseconds. Anyway, it's just a idea and I'm still thinking about it. Feedback welcome :)

**Comment by rvben** [comment]
Hi @n8behavior,

Thanks for the suggestion, I hadn't thought of that. That sounds like a good idea.
An event stream approach seems like the better option for something like this.

Kind regards,

Ruben

**Comment by n8behavior** [comment]
Cool. Okay, then if you don't mind I'll poke around with this and see where it goes. I just didn't want to presume to know your project and how you would like things :)

Tx for sharing your work. Should be fun!
